cmd/wal/scan/cmd.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2024 StreamNative, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scan
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/streamnative/oxia/proto"
+	"log/slog"
+	"math"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/streamnative/oxia/cmd/wal/common"
+	"github.com/streamnative/oxia/server/wal"
+)
+
+type scanOptions struct {
+	json bool
+	//lastEntry      bool
+	//safePointEntry int64
+}
+
+var (
+	options = scanOptions{}
+	Cmd     = &cobra.Command{
+		Use:   "scan",
+		Short: "scan the WAL",
+		Long:  `scan the WAL`,
+		RunE:  exec,
+	}
+)
+
+func init() {
+	// operations
+	//Cmd.Flags().Int64Var(&options.safePointEntry, "last-safe-entry", math.MaxInt64,
+	//	"removes entries from the end of the log that have an offset greater than last safe entry")
+	Cmd.Flags().BoolVar(&options.json, "json", false, "dump entries in json format")
+	//
+	//Cmd.MarkFlagsMutuallyExclusive("last-safe-entry", "truncate-last-entry")
+}
+
+type LogEntry struct {
+	Term          int64                `json:"term"`
+	Offset        int64                `json:"offset"`
+	Timestamp     time.Time            `json:"time"`
+	WriteRequests *proto.WriteRequests `json:"writeRequests"`
+}
+
+func exec(*cobra.Command, []string) error {
+	factory := wal.NewWalFactory(&wal.FactoryOptions{
+		BaseWalDir:  common.WalOption.WalDir,
+		Retention:   math.MaxInt64,
+		SegmentSize: wal.DefaultFactoryOptions.SegmentSize,
+		SyncData:    false,
+	})
+
+	wal, err := factory.NewWal(common.WalOption.Namespace, common.WalOption.Shard, nil)
+	if err != nil {
+		return err
+	}
+	defer wal.Close()
+
+	reader, err := wal.NewReader(wal.FirstOffset() - 1)
```

Comment:
Let's give user a option to set start offset.  :)

Reply:
Done